pkg/logql/log/parser.go

```diff
@@ -259,3 +260,43 @@ func (l *LogfmtParser) Process(line []byte, lbs *LabelsBuilder) ([]byte, bool) {
 }
 
 func (l *LogfmtParser) RequiredLabelNames() []string { return []string{} }
+
+type JSONExpressionParser struct {
+	expressions map[string][]interface{}
+}
+
+func NewJSONExpressionParser(expressions []JSONExpression) (*JSONExpressionParser, error) {
+	var paths = make(map[string][]interface{})
+
+	for _, exp := range expressions {
+		path, err := jsonexpr.Parse(exp.Expression, false)
+		if err != nil {
+			return nil, fmt.Errorf("cannot parse expression [%s]: %w", exp.Expression, err)
+		}
+
+		paths[exp.Identifier] = path
+	}
+
+	return &JSONExpressionParser{
+		expressions: paths,
+	}, nil
+}
+
+func (j *JSONExpressionParser) Process(line []byte, lbs *LabelsBuilder) ([]byte, bool) {
+	if lbs.ParserLabelHints().NoLabels() {
+		return line, true
+	}
+
+	for identifier, paths := range j.expressions {
+		result := jsoniter.ConfigFastest.Get(line, paths...).ToString()
+		if result == "" {
```

Comment:
We can't because this is in a loop of multiple expressions - one expression may not evaluate to a value, but a subsequent one could.

Any expression that doesn't return a value will be ignored. If any expression is evaluated against an empty JSON document, it'll return `InvalidType` which will be stringed to `""`.

Could you expand a bit on what you mean by this?
> Should we implicitly filter these? Is it a best-effort label setting?

Reply:
The current JSON parser will set a `JSONParserErr` here: https://github.com/grafana/loki/blob/master/pkg/logql/log/parser.go#L54. 
I expect we should do something similar here if one of the _expected_ labels does not exist, because we're being explicit.

My original comment was trying to say "if one of the explicitly extracted labels doesn't exist, should we drop the line from this result set?", or _Does this new json extension imply a filter operation_?

Reply:
OK, thanks for clarifying. I think we need to distinguish here between incorrect and malformed expressions.
For example, with `label_format` if you use an **incorrect** expression (i.e. one that doesn't return a value), it will evaluate correctly (blank result).

![image](https://user-images.githubusercontent.com/373762/106849906-38565980-66bc-11eb-8a7f-f308825ec6b5.png)


A **malformed** expression (i.e. one that breaks syntactic rules) should indeed raise an error:
![image](https://user-images.githubusercontent.com/373762/106850034-6fc50600-66bc-11eb-8367-a2879f3d91ad.png)

You're right though that this PR doesn't set a `__error__="JSONParserErr"` label, I'll fix that.

How do you think we should handle malformed/incorrect expressions though?
I'm thinking we follow the behaviour of `label_format`: for malformed queries, raise an error that fails the whole query, and for incorrect queries just set a blank label. We don't, AFAIK, have a mechanism to mark a particular label as being problematic, and since we could be returning multiple labels here this gets a bit tricky in terms of UX.

Reply:
Let's keep parity with what we have & set empty labels when they don't have corresponding JSON fields. i.e. extracting `| json foo="foo"` from `{"bar": 2}` should just yield a `foo=""` label.

Reply:
Done :+1: